chain/client/src/sync/adapter.rs

```diff
@@ -108,9 +109,14 @@ impl SyncAdapter {
 }
 
 /// Interface for network
-#[async_trait::async_trait]
 impl NetworkStateSync for SyncAdapter {
-    async fn send(&self, shard_uid: ShardUId, msg: StateSyncResponse) {
-        self.send(shard_uid, msg.with_span_context());
+    fn send(&self, info: StateResponseInfo) {
+        if let Some(&shard_uid) =
+            self.actor_handler_map.keys().find(|shard_uid| shard_uid.shard_id() == info.shard_id())
```

Comment:
Oh, that won't scale well to tracking 1M shards. :)
Comment which assumptions make O(N) accesses acceptable.